src/Microsoft.AspNet.Mvc.TagHelpers/Internal/ModeMatchResult.cs

```diff
@@ -57,17 +55,39 @@ public class ModeMatchResult<TMode>
                 throw new ArgumentNullException(nameof(tagHelper));
             }
 
-            if (logger.IsEnabled(LogLevel.Warning) && PartiallyMatchedAttributes.Any())
+            if (logger.IsEnabled(LogLevel.Warning) && PartiallyMatchedAttributes.Count > 0)
             {
                 // Build the list of partial matches that contain attributes not appearing in at least one full match
-                var partialOnlyMatches = PartialMatches.Where(
-                    match => match.PresentAttributes.Any(
-                        attribute => PartiallyMatchedAttributes.Contains(
-                            attribute, StringComparer.OrdinalIgnoreCase)));
-                logger.TagHelperPartialMatches<TMode>(uniqueId, viewPath, partialOnlyMatches);
+                var partialOnlyMatches = new List<ModeMatchAttributes<TMode>>();
+                for (var i = 0; i < PartialMatches.Count; i++)
+                {
+                    var presentAttributes = PartialMatches[i].PresentAttributes;
+                    for (var j = 0; j < presentAttributes.Count; j++)
+                    {
+                        var present = presentAttributes[j];
+                        var presentIsPartialOnlyMatch = false;
+                        for (var k = 0; k < PartiallyMatchedAttributes.Count; k++)
+                        {
+                            var partiallyMatched = PartiallyMatchedAttributes[k];
+                            if (string.Equals(partiallyMatched, present, StringComparison.OrdinalIgnoreCase))
+                            {
+                                presentIsPartialOnlyMatch = true;
+                                break;
+                            }
+                        }
+
+                        if (presentIsPartialOnlyMatch)
+                        {
+                            partialOnlyMatches.Add(PartialMatches[i]);
+                            break;
+                        }
+                    }
+                }
```

Comment:
This is some complex machinery for logging. Could we log a bit less information and save 20 lines?